### PR TITLE
Adding hook for custom post status with tests

### PR DESCRIPTION
**Pull request by elysium001**
**Pull Request: Add `msm_sitemap_post_status` WordPress Hook**

**Description:** This pull request introduces a new WordPress hook, `msm_sitemap_post_status`, to facilitate customization of post statuses for inclusion in the sitemap. This feature was specifically requested by a customer to enhance the flexibility of handling custom post statuses in the sitemap functionality.

**Changes Introduced:**

- Added `msm_sitemap_post_status` hook in `msm-sitemap.php`.
- Documented usage of the new hook in the codebase and readme
- Added unit tests to make sure other query and helper functions still work accordingly.

**Background:** A customer expressed the need for a way to use a post status other than 'published' to build the sitemap, prompting the development of this feature. The introduction of the `msm_sitemap_post_status` hook allows developers to extend the sitemap functionality easily and integrate their custom post statuses.

**Related Issues:** 
- internal customer discussion and [P2](https://salesforcep2.wordpress.com/2023/11/21/notes-msm-sitemaps-troubleshooting/)

**Reviewers:** TBD

### CODE REVIEW

File: tests/test-sitemap-functions.php [hunk 1]
```diff
@@ -273,4 +344,58 @@ function test_get_post_ids_for_date( $sitemap_date, $limit, $expected_count ) {
 
 	}
 
+	/**
+	 * Verify get_post_ids_for_date returns expected value with custom status hook
+	 *
+	 * @dataProvider postIdsForDateDataProvider
+	 * @param str $sitemap_date Date in Y-M-D format.
+	 * @param str $limit max number of posts to return.
+	 * @param int $expected_count Number of posts expected to be returned.
```
Comment (GaryJones) [inline tests/test-sitemap-functions.php:353]:
```suggestion
	 * @param string  $sitemap_date   Date in Y-M-D format.
	 * @param string  $limit          Max number of posts to return.
	 * @param int     $expected_count Number of posts expected to be returned.
```

File: tests/test-sitemap-functions.php [hunk 2]
```diff
@@ -273,4 +344,58 @@ function test_get_post_ids_for_date( $sitemap_date, $limit, $expected_count ) {
...
+		$this->assertEquals( array_slice( $created_post_ids, 0, $limit ), $post_ids );
+
+		// remove filter.
+		remove_filter( 'msm_sitemap_post_status', array( $this, 'add_post_status_to_msm_sitemap' ) );
+	}
+
+	/**
+	 * Verify msm_sitemap_post_status filter returns expected value
+	 */
+	 function test_get_post_status() {
```
Comment (GaryJones) [inline tests/test-sitemap-functions.php:384]:
```suggestion
	function test_get_post_status() {
```

File: tests/test-sitemap-functions.php [hunk 3]
```diff
@@ -273,4 +344,58 @@ function test_get_post_ids_for_date( $sitemap_date, $limit, $expected_count ) {
...
+		// verify default value.
+		$this->assertEquals( 'publish', Metro_Sitemap::get_post_status() );
+
+		// add filter and verify value.
+		add_filter( 'msm_sitemap_post_status', array( $this, 'add_post_status_to_msm_sitemap' ) );
+		$this->assertEquals( 'live', Metro_Sitemap::get_post_status() );
+		
+		// remove filter.
+		remove_filter( 'msm_sitemap_post_status', array( $this, 'add_post_status_to_msm_sitemap' ) );
+	 }
```
Comment (GaryJones) [inline tests/test-sitemap-functions.php:395]:
```suggestion
	}
```

File: msm-sitemap.php [hunk 4]
```diff
@@ -323,14 +323,25 @@ public static function disable_canonical_redirects_for_sitemap_xml( $redirect_ur
 		return $redirect_url;
 	}
 
+	/**
+	 * Helper method to get the custom post status.
+	 * Hook allows developers to extend the sitemap functionality easily and integrate their custom post statuses.
+	 *
+	 * @return string
+	 */
+	public static function get_post_status() {
```
[the remark this answers is not in the document]
Comment (GaryJones) [reply msm-sitemap.php:332]: Does this method need to be `public`?

**Review by GaryJones — CHANGES_REQUESTED** [review]

File: msm-sitemap.php [hunk 5]
```diff
@@ -323,14 +323,25 @@ public static function disable_canonical_redirects_for_sitemap_xml( $redirect_ur
...
+		return apply_filters('msm_sitemap_post_status', 'publish');
```
[the remark this answers is not in the document]
Comment (elysium001) [reply msm-sitemap.php]: added checks against get_post_stati values and hook properly documented.

File: msm-sitemap.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (elysium001) [reply msm-sitemap.php]: while it's not required, it does make it much easier to test such as the get_post_stati values mechanism.

File: tests/test-sitemap-functions.php [hunk 6]
```diff
@@ -229,6 +266,40 @@ function test_date_range_has_posts( $start_date, $end_date, $has_post ) {
...
+	 *
+	 * @dataProvider dateRangeHasPostsCustomStatusDataProvider
+	 * @param Str     $start_date Start Date of Range in Y-M-D format.
+	 * @param Str     $end_date  End Date of Range in Y-M-D format.
+	 * @param boolean $has_post Does Range have Post.
+	 */
+	function test_date_range_has_posts_custom_status( $start_date, $end_date, $has_post ) {
+
+		// set msm_sitemap_post_status filter to custom_status.
+		add_filter( 'msm_sitemap_post_status', array( $this, 'add_post_status_to_msm_sitemap' ) );
```
[the remark this answers is not in the document]
Comment (elysium001) [reply tests/test-sitemap-functions.php]: I hear that. I took a stab at this but kept running into a strange issue with anonymous functions not respecting `remove_filter` in the tests. I ended up refactoring slightly to DRY these calls though.

**Review by GaryJones — APPROVED** [review]

**Pull request merged**